app/scripts/background.js

```diff
@@ -184,10 +185,38 @@ let connectExternal;
 browser.runtime.onConnect.addListener(async (...args) => {
   // Queue up connection attempts here, waiting until after initialization
   await isInitialized;
+  const remotePort = args[0];
+  const { sender, name } = remotePort;
 
-  // This is set in `setupController`, which is called as part of initialization
-  connectRemote(...args);
+  const url = sender?.url;
+  const detectedProcessName = url ? getEnvironmentType(url) : '';
+  const processName = name;
+
+  if (
+    processName === 'metamask-contentscript' &&
+    detectedProcessName === 'background'
+  ) {
+    // This is set in `setupController`, which is called as part of initialization
+    connectRemote(...args);
+    return;
+  }
+
+  if (detectedProcessName !== processName) {
+    throw new Error('UNEXPECTED_CONNECTION_ATTEMPT');
+  }
+
+  const connectionId = generateConnectionId(remotePort, processName);
+  const connection = openMetamaskConnections.get(connectionId);
+
+  if (Boolean(connection) === false) {
+    // This is set in `setupController`, which is called as part of initialization
+    connectRemote(...args);
```

Comment:
Nit: The flow of execution of this method has been complicated quite a bit; beforehand `connectRemote` was called once in the main body of the function, but now it's called twice in separate branches.

Perhaps we could rearrange things to put validation first, then call `connectRemote` once at the end?